### PR TITLE
Data integration add frequency plot

### PR DESCRIPTION
**Pull request by cosa65**
# Background
#### Link to issue 
https://biomage.atlassian.net/browse/BIOMAGE-524?atlOrigin=eyJpIjoiMGViZTZmMTg0YjRlNGYwMThmZTk1Mzc2YTY3ODk1ZTUiLCJwIjoiaiJ9

#### Link to staging deployment URL 
https://ui-nj189pnipdy3jr8ke6z7rbwgny.scp-staging.biomage.net/

#### Anything else the reviewers should know about the changes here

# Changes
### Code changes

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [x] Unit tests written
- [x] Tested locally with Inframock
- [x] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [ ] Relevant Github READMEs updated
- [ ] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by vickymorrison (this is her username, tag her if you need approval)

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [ ] Photo of a cute animal attached to this PR

### CODE REVIEW

File: src/pages/experiments/[experimentId]/data-processing/data-integration/components/DataIntegration.jsx [hunk 1]
```diff
@@ -21,31 +18,62 @@ import LegendEditor from '../../../plots-and-tables/components/LegendEditor';
...
-  transGrid: 10,
-  width: 530,
-  height: 400,
-  maxWidth: 720,
-  maxHeight: 530,
+import isBrowser from '../../../../../../utils/environment';
+
+import loadCellSets from '../../../../../../redux/actions/cellSets/loadCellSets';
+
+// TODO This loadPlotConfig should probably be changed on the redux ticket? because it fetches in a url that seems dedicated to plots and tables:
```
Comment (ivababukova) [inline src/pages/experiments/[experimentId]/data-processing/data-integration/components/DataIntegration.jsx]: is this for the styling settings?

File: src/pages/experiments/[experimentId]/data-processing/data-integration/components/DataIntegration.jsx [hunk 2]
```diff
@@ -66,33 +94,64 @@ const defaultPlotStylingConfig = {
...
+    frequencyPlot: useSelector((state) => state.componentConfig[frequencyPlotConfigRedux.uuid]?.config),
+    elbowPlot: _.cloneDeep(defaultElbowPlotStylingConfig),
+  };
+
+  const [activePlotKey, setActivePlotKey] = useState('frequencyPlot');
+  const config = persistedConfigs[activePlotKey];
+
+  const setCurrentConfig = () => {
+    // This will be used for dispatching the config updates to redux
+  };
```
Comment (ivababukova) [inline src/pages/experiments/[experimentId]/data-processing/data-integration/components/DataIntegration.jsx]: if it won't be used for these changes, I suggest that you remove it -- we can add it during the work on redux

File: src/pages/experiments/[experimentId]/data-processing/data-integration/components/DataIntegration.jsx (hunk 1 above)
Comment (cosa65) [reply src/pages/experiments/[experimentId]/data-processing/data-integration/components/DataIntegration.jsx]: Yes

Comment (cosa65) [reply src/pages/experiments/[experimentId]/data-processing/data-integration/components/DataIntegration.jsx]: Im going to remove this comment and add it to the ticket.

File: src/pages/experiments/[experimentId]/data-processing/data-integration/components/fake_new_data.json [hunk 3]
```diff
@@ -1 +1,15 @@
...
...ariance": 0.01687049739597}, {"PC": 28, "percentVariance": 0.0168331069010516}, {"PC": 29, "percentVariance": 0.0167849937299379}, {"PC": 30, "percentVariance": 0.0167707377515886}, {"PC": 31, "percentVariance": 0.0167222909834068}, {"PC": 32, "percentVariance": 0.0166360384033748}, {"PC": 33, "percentVariance": 0.0165978110924039}, {"PC": 34, "percentVariance": 0.0165756885803951}, {"PC": 35, "percentVariance": 0.0165443774624861}, {"PC": 36, "percentVariance": 0.0164812140708958}, {"PC": 37, "percentVariance": 0.0164573941892336}, {"PC": 38, "percentVariance": 0.0164319809163447}, {"PC": 39, "percentVariance": 0.0163884779595168}, {"PC": 40, "percentVariance": 0.0163647208511092}, {"PC": 41, "percentVariance": 0.0163081952112513}, {"PC": 42, "percentVariance": 0.016301594442556}, {"PC": 43, "percentVariance": 0.0162582683913705}, {"PC": 44, "percentVariance": 0.0161894672132257}, {"PC": 45, "percentVariance": 0.0161749043236022}, {"PC": 46, "percentVariance": 0.0161602173615978}, {"PC": 47, "percentVariance": 0.0161414499083575}, {"PC": 48, "percentVariance": 0.0161041148570463}, {"PC": 49, "percentVariance": 0.0160692972009101}, {"PC": 50, "percentVariance": 0.0160396861269541}]
```
Comment (StefanBabukov) [inline src/pages/experiments/[experimentId]/data-processing/data-integration/components/fake_new_data.json]: Do we need to have this file commited?

File: src/utils/sharedPlots/FrequencyPlot.jsx [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+      metadataClusters.forEach((metadataCluster) => {
+        const metadataIds = Array.from(properties[metadataCluster.key].cellIds);
+
+        // const exhangedValues = [0, 21, 29, 31, 34, 55, 58];
+        // // 206 198
+        // if (metadataIds.count === 206) {
+        //   metadataIds = metadataIds.filter((value) => !exhangedValues.contains(value));
+        // } else {
+        //   metadataIds = metadataIds.concat(exhangedValues);
+        // }
```
Comment (StefanBabukov) [inline src/utils/sharedPlots/FrequencyPlot.jsx]: Left comments

File: src/pages/experiments/[experimentId]/data-processing/data-integration/components/DataIntegration.jsx [hunk 5]
```diff
@@ -24,31 +16,62 @@ import LegendEditor from '../../../plots-and-tables/components/LegendEditor';
...
+import loadCellSets from '../../../../../../redux/actions/cellSets/loadCellSets';
+
+import { loadProcessingSettings } from '../../../../../../redux/actions/experimentSettings';
+
+import {
+  loadPlotConfig,
+} from '../../../../../../redux/actions/componentConfig/index';
+
+import CalculationConfig from './CalculationConfig';
+import fakeData from './fake_new_data.json';
```
Comment (StefanBabukov) [inline src/pages/experiments/[experimentId]/data-processing/data-integration/components/DataIntegration.jsx]: Are we using fake data for now(Also addressed in the first comment)?

**Review by StefanBabukov — DISMISSED** [review]

File: src/pages/experiments/[experimentId]/data-processing/data-integration/components/DataIntegration.jsx (hunk 5 above)
Comment (cosa65) [reply src/pages/experiments/[experimentId]/data-processing/data-integration/components/DataIntegration.jsx]: It was being used from before too, we need to use fake data for the elbow plot because it displays information about a the result of pca (which is not implemented in the worker yet)

Comment (StefanBabukov) [reply src/pages/experiments/[experimentId]/data-processing/data-integration/components/DataIntegration.jsx]: Right, thanks for the explanation!

File: src/utils/sharedPlots/FrequencyPlot.jsx [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+              type: 'stack',
+              groupby: ['x'],
+              sort: { field: 'c' },
+              field: 'y',
+            },
+          ],
+        },
+      ],
+
+      scales: [
```
Comment (ivababukova) [inline src/utils/sharedPlots/FrequencyPlot.jsx]: why did you move the plot Vega spec in here?

Comment (cosa65) [reply src/utils/sharedPlots/FrequencyPlot.jsx]: To follow the logic used in ElbowPlot and the plots in configure embedding, where plots are in separate components.

If you are asking about why moving them to utils, then because since this plot is being used in different places then I felt it made more sense to have them in a shared space.

Comment (cosa65) [reply src/utils/sharedPlots/FrequencyPlot.jsx]: Moved them to sharedPlots in /components.

File: src/components/plots/FrequencyPlot.jsx [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+    let data = [];
+
+    const proportionClusters = hierarchy.filter((cluster) => (
+      cluster.key === config.proportionGrouping))[0]?.children;
+
+    if (!proportionClusters) {
+      return [];
+    }
+
+    const clustersInXAxis = getClustersInXAxis(config.xAxisGrouping);
```
Comment (ivababukova) [inline src/components/plots/FrequencyPlot.jsx]: are there any unit tests for this plot? If not, we should add some, especially for this `getClustersInXAxis` function, which is full of logic

File: src/pages/experiments/[experimentId]/data-processing/data-integration/components/plots/ElbowPlot.jsx [hunk 8]
```diff
@@ -0,0 +1,133 @@
+import React from 'react';
+import { Vega } from 'react-vega';
+import PropTypes from 'prop-types';
+
+const ElbowPlot = (props) => {
```
Comment (ivababukova) [inline src/pages/experiments/[experimentId]/data-processing/data-integration/components/plots/ElbowPlot.jsx]: shouldn't this be stored under ` src/components/plots`, next to the Frequency plot?

File: src/redux/reducers/componentConfig/initialState.js [hunk 9]
```diff
@@ -296,6 +298,34 @@ const interactiveHeatmapInitialConfig = {
   legendIsVisible: true,
 };
 
+// DATA INTEGRATION - Frequency
+const dataIntegrationFrequencyInitialConfig = {
```
Comment (ivababukova) [inline src/redux/reducers/componentConfig/initialState.js:302]: many of the fields in `dataIntegrationFrequencyInitialConfig` have values that are the same as the values for `frequencyInitialConfig`. To make this easier to change and more clear, what you can do is make `dataIntegrationFrequencyInitialConfig`  be based on `frequencyInitialConfig` and then just specify the fields that differ. Something like this:

```
const dataIntegrationFrequencyInitialConfig = {
    ...frequencyInitialConfig,
    proportionGrouping: 'sample',
    and so on
}
```

**Comment by ivababukova** [comment]
I just took a look at staging and I noticed that the plot is not showing up, see the screenshot:

Screenshot 2021-02-15 at 09 13 06

Is this expected behaviour?

**Comment by cosa65** [comment]
> I just took a look at staging and I noticed that the plot is not showing up, see the screenshot:
> 
> Screenshot 2021-02-15 at 09 13 06
> 
> Is this expected behaviour?

It is looking for the 'Sample' grouping which is not added in the default experiment, if you use the experiment 'multisample' you'll see it is working ok @ivababukova 

https://ui-nj189pnipdy3jr8ke6z7rbwgny.scp-staging.biomage.net/experiments/multisample/data-exploration

File: src/pages/experiments/[experimentId]/data-processing/data-integration/components/plots/ElbowPlot.jsx (hunk 8 above)
Comment (cosa65) [reply src/pages/experiments/[experimentId]/data-processing/data-integration/components/plots/ElbowPlot.jsx]: I didn't store it there because it wasn't shared anywhere else in the app, but if you believe it makes more sense to have it there I can do that

**Comment by ivababukova** [comment]
Yup, just checked it, looks good! Here is the link: https://ui-nj189pnipdy3jr8ke6z7rbwgny.scp-staging.biomage.net/experiments/multisample/data-processing

**Review by ivababukova — APPROVED** [review]
